### DISCUSSION TITLE
Adding sample projects for extensions

### DISCUSSION THREAD

**Discussion by sfederici** [TurboWarp/extensions#915] [Ideas]
Extensions are certainly cool, but without a detailed help it is difficult sometime to imagine how to use them to get what you want. It would be helpful adding sample projects under each extension in the extension gallery page https://extensions.turbowarp.org/

**Comment by CubesterYT** [comment]
This is a great idea!

**Comment by DNin01** [comment]
I would be willing to help out with this! It would be an even better way to teach people how to use the extensions, without making creators follow the documentation on their own.

Just need a good way to make them easy to open and run. The packager wouldn't be ideal because you wouldn't be able to enter the editor, I don't think. I share TurboWarp projects at https://github.com/DNin01/Scratch-projects but it would be better if no download was required.

**Comment by GarboMuffin** [comment]
We have been slowly moving in this direction by adding docs pages like this one - https://extensions.turbowarp.org/box2d (most importantly, I've made these actually *enjoyable* to write, so more should be popping up)

Full sample projects would indeed be a good improvement here

If someone writes a few, I can make a folder for them and start putting links in the UI for them

**Reply by LilyMakesThings** [reply]
On it

**Reply by CubesterYT** [reply]
I've actually made one a while ago, I use it personally to test Window Controls. I'll see what I can do to clean it up to make it an example.

**Comment by DNin01** [comment]
@GarboMuffin Do you have a certain format you want the sample projects in? Should they be tutorials or just code examples? Or does it even matter?

And how should they be stored? Would each sample project simply be an .sb3 file or would we need to do some sort of packaging?

**Reply by GarboMuffin** [reply]
We would just do the .sb3. Being able to easily see the source code would be important for this.

We already have a tool for this - ?project_url=

**Reply by sfederici** [reply]
Yes, I agree. I’m going to review the projects as soon as I’m back to my PC. It could be good to have comments attached to the extension blocks used in the projects

**Reply by DNin01** [reply]
I would assume you can just host the .sb3 files on GitHub Pages and then ?project_url= will work. Maybe I just don't know how to copy the link correctly, but I tried to load a **raw** link and it didn't work, with this error:
```
Error in GUI [location=https://turbowarp.org/?project_url=https://github.com/DNin01/Scratch-projects/raw/main/turbowarp/tools/meter-and-slider/Meter+and+Slider.sb3]: Error: [tw-progress-monitor] xhr failed
```

**Reply by DNin01** [reply]
Oh yeah, and here's the first error in the browser console:
```
Access to XMLHttpRequest at 'https://github.com/DNin01/Scratch-projects/raw/main/turbowarp/tools/meter-and-slider/Meter%20and%20Slider.sb3' from origin 'https://turbowarp.org' has been blocked by CORS policy: The 'Access-Control-Allow-Origin' header has a value 'https://render.githubusercontent.com' that is not equal to the supplied origin.
```

**Comment by GarboMuffin** [comment]
@sfederici 

@LilyMakesThings sent me some projects (more than just the few below), are these the kinds of projects you're looking for?

https://share.turbowarp.org/projects/02ca2ba8-3363-447f-a4cc-39a7dd770f98

https://share.turbowarp.org/projects/09b6a677-b577-4eb3-a7d2-5673a66122f0

https://share.turbowarp.org/projects/b30f9681-78fa-4b93-96f4-9963ff3751d6

(Placeholder lacks an integrated editor, but you can download them manually. When sample projects are done they will have easy editor access)

**Reply by sfederici** [reply]
Yes, this exactly is the kind of projects I would like to be listed in the extension gallery.

**Comment by DNin01** [comment]
I just finished one for Box2D: https://share.turbowarp.org/projects/fc8302bc-f57c-43c0-8b25-97ee9c0dbc3b `v2.0`

It has five different demonstrations to show a few different things you can do.

**Reply by DNin01** [reply]
Sometimes the Dango falls instead of spinning in place. After making the demonstration auto-repeat, it happens every time it repeats until you switch demonstrations or click the green flag. This must be an extension bug.

I also wanted to use relative positioning but that was buggy too: #935

**Reply by GarboMuffin** [reply]
I feel that it would be best if each of the sample projects are as minimal as possible without fluff like a scene changer in this case. Having more than one sample should be something to consider

**Reply by CubesterYT** [reply]
That would make more sense, it allows for the examples to be more focused.

**Reply by DNin01** [reply]
Alternatively, I could separate each scene into its own sprite.

**Reply by DNin01** [reply]
Updated again, this time with separate sprites for each demo.

**Comment by GarboMuffin** [comment]
A few sample projects are now on https://extensions.turbowarp.org/ (stretch, box2d, cursor, a couple others)

Making more is really simple - just create a file using an extension, save to your computer, and submit a pull request dropping it into the "samples" folder with a unique name. That's it. It'll figure out which extensions it applies to on its own.

**Reply by sfederici** [reply]
I like the addition of the buttons when you hover the extension. I think it is intuitive. What if there is more than one sample? Will the button direct you to a page where you can select the sample?

**Reply by GarboMuffin** [reply]
It'll show a dropdown list

**Reply by DNin01** [reply]
Interesting. If that was implemented, I could separate all the demos I made into their own projects, each focusing on a function or two. The sample code would then be **much** easier to deal with.

**Reply by sfederici** [reply]
Good

**Reply by GarboMuffin** [reply]
It looks like this:

![image](https://github.com/TurboWarp/extensions/assets/33787854/557b0c65-5d18-4d07-8a2c-9a4b43aa8c94)

I will fix the issue of it not being able to show too many items once it becomes a problem

**Reply by sfederici** [reply]
Very good. I like it

**Comment by sfederici** [comment]
> It has five different demonstrations to show a few different things you can do.



> I feel that it would be best if each of the sample projects are as minimal as possible without fluff like a scene changer in this case. Having more than one sample should be something to consider

I like the samples. I agree that, at least, I would separate each samples in a different sprites and move the "navigation" sprite to the last position (this is the "cosmetic" work my students appreciate a lot)

**Comment by DNin01** [comment]
@GarboMuffin I made one for Temporary Variables that might be a little clearer:
https://share.turbowarp.org/projects/54805640-1060-4b72-b9f7-86a3a0e2d0c4

Everything is in the editor. It's simpler than the Box2D one but there are still 2 sprites.

**Comment by DNin01** [comment]
And one for Tween:
https://share.turbowarp.org/projects/c7b6fc0c-a773-47f4-a015-58a5725f7b7d

**Reply by scra-bit** [reply]
I  made an example project for the navigator info extension

https://share.turbowarp.org/projects/68a92520-6b91-44c2-852d-e71e675d8e61

**Comment by scra-bit** [comment]
@GarboMuffin              I  made an example project for the navigator info extension

https://share.turbowarp.org/projects/68a92520-6b91-44c2-852d-e71e675d8e61

**Comment by sfederici** [comment]
> And one for Tween: https://share.turbowarp.org/projects/c7b6fc0c-a773-47f4-a015-58a5725f7b7d

Minimal, but useful I guess

**Comment by Pizzalover4783** [comment]
@GarboMuffin 
Could you give some concrete rules for what should and should not be in a sample project?
I'm currently making one myself, but I'm not sure if you'll allow it.

**Reply by Pizzalover4783** [reply]
https://drive.google.com/file/d/1RxZJ8KfofD2ftQ20h8iNLvIho3cZiuT4/view?usp=sharing
Here's the sb3

**Comment by CubesterYT** [comment]
@Pizzalover4783 You don't need to make a sample for Window Controls, I've already made one, I'm putting up a PR to add it tomorrow. Also the rules aren't really specified, but from what I see on how the sample system works, the only extension that's supposed to be in the sample is the extension you want to sample, as the sampling system unzips the project to check which extension it's sampling.

**Comment by xTvii** [comment]
made a quick little one to demonstate the video extension. simple, maybe unnessecary, but fun
https://share.turbowarp.org/projects/10138760-a2c7-47b3-b14e-88549473a160
